(CDAP-888) Dataset Authorization Part I: Created a service that cache…

bdmogal opened the pull request:
…s the privileges of the user running the program. The privileges are refreshed at a configurable time interval.

In this PR:
1. Split `enforce(EntityId, Principal, Action)` into a separate `AuthorizationEnforcer` interface.
2. `Authorizer` extends `AuthorizationEnforcer`, so external authorizers still only have to implement a single interface - `Authorizer`.
3. Create a `ProgramAuthorizationEnforcementService` which maintains a cache of `Map<Principal, Set<Privilege>`, to be used while accessing datasets from programs.
4. The cache is updated once just before the program is started, so all privileges for the user running the program are cached. From then on, the cache is updated at a configurable time interval, and enforcement is only based on cached policies.

Follow up work:
1. Implement remote-style APIs being introduced in #5934 for listing user privileges, so communication with authorization backends can happen as the 'cdap' user (which would presumably be whitelisted).
2. Test in conjunction with secure impersonation
3. Test with Apache Sentry after 1.
4. Integration with Read/Write dataset APIs
4. Authorization enforcement for dataset REST APIs

Issue: [CDAP-888](https://issues.cask.co/browse/CDAP-888)
Build: http://builds.cask.co/browse/CDAP-DUT4282

Commit:
(CDAP-888) Dataset Authorization Part I: Created a service that caches the privileges of the user running the program. The privileges are refreshed at a configurable time interval.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:
```diff
@@ -297,8 +300,9 @@ public void run() {
     LOG.info("Starting metrics service");
```
Seems like this log message is wrong.

bdmogal replied:
yup. will remove, there's a log for the program 5 lines below.

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,154 @@
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (!cacheEnabled) {
```
can't we simply return also in the case that authorization is disabled?

bdmogal replied:
if authorization is disabled, in `updatePrivileges`, the `Authorizer` will be a `NoOpAuthorizer`, which will just return an empty set for `listPrivileges`. Unsure if we need to consider that condition here as well. I'll think about it.

bdmogal replied:
I tried to use an approach which created multiple `AuthorizationEnforcementService` classes that get injected based on cConf here - https://github.com/caskdata/cdap/commit/96dd11b377d45f0457763c10c2ed01be19ea1d61. However, it just seems to me that at this stage, it adds a lot of almost no-op code. Not sure if there is much value to doing that yet. 

I have however added a check for authorizationEnabled here just to cut the point across, as well as a warning in the `startUp` method if authorization is disabled but caching is enabled.

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,154 @@
+    try {
+      if (authPolicyCache.containsKey(principal)) {
+        Set<Privilege> privileges = authPolicyCache.get(principal);
+        for (Privilege privilege : privileges) {
```
can't we use `Set#contains`

bdmogal replied:
hmmm, there was a problem here, it wasn't checking for the entity. fixed now, also updated test.

anwar6953 commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,152 @@
+    }
+    LOG.debug("Running authorization enforcement service iteration...");
+    // we want to block all access to the cache while the cache update is running
+    writeLock.lock();
```
what happens if you have a lot of threads performing reads - need to make sure that the write lock acquisition isn't blocked for too long

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:
```diff
@@ -356,14 +360,38 @@ public void error(Throwable cause) {
+   * On an authorization-enabled cluster, if caching is enabled too, updates the cache in the
+   * {@link AuthorizationEnforcementService} with the privileges of the user running the program.
+   */
+  private void updateAuthCache() {
```
Why this method belongs to this class? Shouldn't it be part of what the `AuthorizationEnforcementService` do as part of it's startup?

bdmogal replied:
This is only called once, before the program starts up. It is here, to pass the `Principal` running the program from `AbstractProgramTwillRunnable` to `AuthorizationEnforcementService`.

chtyim replied:
I mean why on the start up of the service cannot do this? E.g. it can just get the current UGI and do a first fetch on start up?

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/gateway/handlers/AuthorizationHandlerTest.java:
```diff
@@ -39,6 +39,7 @@
 import co.cask.cdap.security.authorization.AuthorizationContextFactory;
 import co.cask.cdap.security.authorization.AuthorizerInstantiator;
 import co.cask.cdap.security.authorization.DefaultAuthorizationContext;
+import co.cask.cdap.security.authorization.InMemoryAuthorizer;
```
unused import?

bdmogal replied:
no, I moved this class from `cdap-app-fabric/src/main/test` to `cdap-security/src/main/test`. That's why we need the import now.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:
```diff
@@ -454,6 +483,12 @@ public Cancellable announce(String serviceName, int port) {
+          bind(AuthorizationEnforcer.class).to(AuthorizationEnforcementService.class).in(Scopes.SINGLETON);
+          // also bind AuthorizationEnforcementService as a singleton. This binding is used while starting/stopping
+          // the service itself.
+          bind(AuthorizationEnforcementService.class).in(Scopes.SINGLETON);
```
It's better not to expose the implementation. Shouldn't it be the `Authorizer` to be exposed?

chtyim replied:
Seems like it only implements `AuthorizationEnforcer`. In that case, it's better to have two interfaces, `AuthorizationEnforcer` and `AuthorizationEnforcerService`, with the latter one extends from the first one. Then you have an implementation class `DefaultAithroizationEnforcerService` that implements the latter one. You can then have bindings like this:

``` java
bind(AuthorizationEnforcer.class).to(AuthorizationEnforcerService.class);
bind(AuthorizationEnforcerService.class).to(DefaultAuthorizationEnforcerService.class).in(Scopes.SINGLETON);
```

chtyim replied:
By doing this, you don't have to expose the implementation, which is the whole point of guice.

bdmogal replied:
`AuthorizationEnforcementService` has two roles:
1. An `AbstractScheduledService` that refreshes privileges periodically. A handle to this is obtained using `injector.getInstance(AuthorizationEnforcementService.class)`.
2. An `AuthorizationEnforcer`, which is used by the dataset methods enforce authorization. Presumably, a handle to this will be obtained using `injector.getInstance(AuthorizationEnforcer.class)`.

Is there a better way to do this?

Also, I didn't want to make `AuthorizationEnforcementService` an `Authorizer`, because `Authorizer` can modify privileges as well, and I didn't want to expose that functionality in `AuthorizationEnforcementService`.

bdmogal replied:
Just saw your response, will use that approach.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,159 @@
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheTtlSeconds, TimeUnit.SECONDS);
```
So the TTL is not exactly a TTL. This is because it takes time to update the cache and the next update time is actually the previous update finish time + cacheTTL. If we want to call it ttl, we should have
1. Start updating the cache earlier then the TTL
2. Have the cache expires automatically when TTL is reached

bdmogal replied:
Would it help if we cleared the cache as the first step in `runOneIteration`? It would have some overhead of clearing the map, but it shouldn't be as much as refreshing the entire cache?

chtyim replied:
But then would it block all the lookup threads?

bdmogal replied:
That's true. How about using guava's `Cache`, with `expireAfterWrite`?

bdmogal replied:
However, with that, it would mean that there could be a small interval between the time when  an entry expires and the time that the refresh thread completes, when programs would get `UnauthorizedException`. That should be ok?

bdmogal replied:
Updated to use `LoadingCache` per discussion

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,159 @@
+  private final boolean authorizationEnabled;
+  private final boolean cacheEnabled;
+  private final int cacheTtlSeconds;
+  private final Map<Principal, Set<Privilege>> authPolicyCache;
```
Is this service supposed to work with multiple `Principal`? Or should it only be used in program runtime container, which there should only be one principal?

bdmogal replied:
The idea was to make it extensible so it can be functional with multiple `Principals` too. This was from a discussion earlier where we decided we could use potentially use it in other places (streams?).

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+    writeLock.lock();
+    try {
+      for (Principal principal : authPolicyCache.asMap().keySet()) {
+        updatePrivileges(principal);
```
You might want to log a debug message about which principal is getting updated.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+    }
+    LOG.debug("Running authorization enforcement service iteration...");
+    // we want to block all access to the cache while the cache update is running
+    writeLock.lock();
```
You don't need the lock anymore. The `LoadingCache` is thread safe.

bdmogal replied:
yes 😞 . fixed.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+      return;
+    }
+    Authorizer authorizer = authorizerInstantiator.get();
+    Set<Privilege> privileges = authorizer.listPrivileges(principal);
```
instead of duplicating the logic, you can just do a `cache.get(principal)`.

bdmogal replied:
but that will not update a principal's privileges if the entry in the cache has not expired, right? This method is meant to always update the cache entry, no matter if the entry is expired or not.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+
+  @Override
+  protected void startUp() throws Exception {
+    super.startUp();
```
No need to call super.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+    readLock.lock();
+    try {
+      Set<Privilege> privileges = authPolicyCache.get(principal);
+      if (privileges.contains(new Privilege(entity, action))) {
```
The logic would be more direct if you have

``` java
if (!privileges.contains(...)) {
  throw exception;
}
```

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,155 @@
+
+  @Override
+  protected void shutDown() throws Exception {
+    super.shutDown();
```
no need to call super. Also I think it's better to call `cache.invalidAll`

bdmogal commented:
@chtyim addressed all comments, other than a slight difference in https://github.com/caskdata/cdap/pull/5972#discussion_r70304556, the `AuthorizationEnforcementService` cannot be an interface because it needs to extend `AbstractScheduledService`.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,31 @@
+ * {@link Constants.Security.Authorization#CACHE_ENABLED}, authorization policies are cached locally.
+ * A thread refreshes the cached policies periodically.
+ */
+public abstract class AuthorizationEnforcementService extends AbstractScheduledService
```
Why this has to be an abstract class? Can't it be an interface extends from `Service` and `AuthorizationEnforcer`?

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  private Set<Privilege> getPrivileges(Principal principal) throws Exception {
+    return authorizerInstantiator.get().listPrivileges(principal);
```
This method should throw exception if this service is not running.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,163 @@
+    try {
+      userName = UserGroupInformation.getCurrentUser().getShortUserName();
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
Why propagate this one while just log the other one (down below)?

bdmogal replied:
My thinking was:
1. This is done before the program starts running. Say due to a transient error, this operation fails.
2. When the program starts, and again requests for privileges from the cache, the LoadingCache does not have anything, so it makes a call to the authorization backend, and if the transient error is fixed before this first access from the program occurs, the program could still run fine. If it doesn't get fixed, then the user gets an error during program runtime, if he tries to execute an operation that requires authorization.

However, for this case, the scant javadoc for `getCurrentUser()` says that it throws `IOException` if login fails. That didn't seem like a transient error.

Do you think that overcomplicates it, and we should just standardize on failing the program if privileges can't be updated before it starts? If the above approach is ok, I can add a comment to explain that.

chtyim commented:
only a few questions/comments left

bdmogal commented:
@chtyim addressed comments. Please review again.

chtyim commented on cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java:
```diff
@@ -0,0 +1,180 @@
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    this.cacheRefreshIntervalSecs = cConf.getInt(Constants.Security.Authorization.CACHE_REFRESH_INTERVAL_SECS);
```
You might want to validate that the interval and ttl are > 0. Also log a warning if ttl <= interval.

chtyim commented:
Just one comment left about the configuration validation. The rest LGTM.

bdmogal commented:
Addressed final comment as well, which was very minor. Not going to run build again, ran relevant tests and checkstyle locally. Squashed commits. Build: http://builds.cask.co/browse/CDAP-DUT4282-20. Tested on cluster. Merging, to unblock work on upcoming PRs.
